system-addon/lib/HighlightsFeed.jsm

```diff
@@ -93,46 +102,51 @@ this.HighlightsFeed = class HighlightsFeed {
 
       // If we already have the image for the card, use that immediately. Else
       // asynchronously fetch the image.
-      const image = currentImages[page.url];
-      if (!image) {
-        this.fetchImage(page.url, page.preview_image_url);
+      if (!page.image) {
+        this.fetchImage(page);
       }
 
       // We want the page, so update various fields for UI
       Object.assign(page, {
-        image,
         hasImage: true, // We always have an image - fall back to a screenshot
         hostname,
         type: page.bookmarkGuid ? "bookmark" : page.type
       });
 
       // Add the "bookmark" or not-skipped "history"
-      this.highlights.push(page);
+      highlights.push(page);
       hosts.add(hostname);
 
       // Skip the rest if we have enough items
-      if (this.highlights.length === HIGHLIGHTS_MAX_LENGTH) {
+      if (highlights.length === HIGHLIGHTS_MAX_LENGTH) {
         break;
       }
     }
 
-    SectionsManager.updateSection(SECTION_ID, {rows: this.highlights}, this.highlightsLastUpdated === 0 || broadcast);
+    SectionsManager.updateSection(SECTION_ID, {rows: highlights}, broadcast);
     this.highlightsLastUpdated = Date.now();
+    this.highlightsLength = highlights.length;
   }
 
   /**
    * Fetch an image for a given highlight and update the card with it. If no
-   * image is available then fallback to fetching a screenshot. Update the card
-   * in `this.highlights` so that the image is cached for the next refresh.
+   * image is available then fallback to fetching a screenshot.
    */
-  async fetchImage(url, imageUrl) {
-    const image = await Screenshots.getScreenshotForURL(imageUrl || url);
-    SectionsManager.updateSectionCard(SECTION_ID, url, {image}, true);
+  async fetchImage(page) {
+    // Request a screenshot if we don't already have one pending
+    const {preview_image_url: imageUrl, url} = page;
+    if (!page.fetchingImage) {
+      page.fetchingImage = new Promise(async resolve => {
```

Comment:
Does  `.fetchingImage` need to be removed on completion? I see it sticking around in the content reducers

Reply:
Shared this adding and removing logic for `__fetchingScreenshot` as part of `Screenshots.maybeGetAndSetScreenshot` and updated `__updateCache` to delete properties.